letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +75,123 @@ def validate_domain(self, domain):
+            for challenge in authorizations.body.challenges:
+                # we might be given multiple challenges,
+                # we only need to respond to one.
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP:
+                    self._respond_challenge_http(challenge, account_key)
+                    break
+                elif challenge.chall.typ == TYPE_CHALLENGE_DNS:
+                    self._respond_challenge_dns(challenge, domain)
+                    break
+                else:
+                    _logger.error(
+                        "Unknown challenge received from letsencrypt.")
+        # let them know we are done and they should check
+        from datetime import datetime, timedelta
+        poll_deadline = datetime.now() + timedelta(minutes=5)
+        order_resource = client.finalize_order(authzr, poll_deadline)
+        with open(join(_get_data_dir(), '%s.crt' % domain), 'w') as crt:
+            crt.write(order_resource.fullchain_pem)
+            _logger.info('SUCCESS: Certificate saved :%s', crt.name)
+            ir_config_parameter = self.env['ir.config_parameter']
+            reload_cmd = ir_config_parameter.get_param(
+                'letsencrypt.reload_command', False)
+            if reload_cmd:
+                self.call_cmdline(['sh', '-c', reload_cmd])
+            else:
+                _logger.warning("No reload command defined.")
+
+    def _create_client(self, account_key):
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('DRY_RUN'):
+            directory_url = V2_STAGING_DIRECTORY_URL
         else:
-            _logger.info('no command defined for reloading webserver, please '
-                         'do it manually in order to apply new certificate')
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        return client.ClientV2(directory_json, net)
+
+    def _make_csr(self, account_key, domain_key_file, domain):
+        parameter_model = self.env['ir.config_parameter']
+        altnames = parameter_model.search(
+            [('key', 'like', 'letsencrypt.altname.%')],
+            order='key')
+        valid_domains = [x.value for x in altnames] + [domain]
+        _logger.info(
+            'Making CSR for the following domains: ' + str(valid_domains))
+        return crypto_util.make_csr(
+            open(domain_key_file).read(), valid_domains)
+
+    def _respond_challenge_http(self, challenge, account_key):
+        """Respond to the HTTP challenge by writing the file to serve.
+        """
+        token = base64.urlsafe_b64encode(challenge.token)
+        challenge_file = join(_get_challenge_dir(), '%s' % token)
+        with open(challenge_file, 'wb') as challenge_file:
+            # remove the trailing equals sign (=)
+            challenge_file.write(token + '.' + base64.urlsafe_b64encode(
+                account_key.thumbprint())[:-1])
```

Comment:
a trainling `=` is base64's padding character, so it's a matter of chance that this happens to be just one in your tests

Reply:
There are two trailing `=` characters. But one of them needs to be removed. https://github.com/certbot/josepy/blob/master/src/josepy/b64.py#L32

Reply:
I wanted to ask you to do exactly that: `rstrip` removes *all* characters. But then why don't you use the function you linked anyways?

As for the amount of characters, check

```python
>>> import base64
>>> base64.b64encode('test')
'dGVzdA=='
>>> base64.b64encode('test1')
'dGVzdDE='
>>> base64.b64encode('test12')
'dGVzdDEy'
```